3.1 release blogpost (#1284)
scala/scala-lang: blog/_posts/2021-10-21-scala-3.1.0-released.md

If you are a library maintainer, *updating to 3.1.0 will force all of your users to update to 3.1.0 as well*. We understand that the current state of binary compatibility may be unsatisfactory. We are actively working on technical solutions to support forward-compatibility in 3.2.0. In the meantime, we recommend testing your library with Scala 3.1.0 and 3.0.2, but _publishing_ it with 3.0.2. This will allow downstream users to use your library even if they can’t update to Scala 3.1.0. You can find an example of build that follows this recommendation [here](https://github.com/typelevel/scalacheck/pull/847).

## What's new in 3.1

### New experimental feature: safer exceptions

A new experimental feature that allows declaring and checking which exceptions can be thrown. It relies on the effects as implicit capabilities pattern.

> You can opt-in using following import.

Reviewer: ```suggestion
You can opt-in using the following import.
```

```scala
import language.experimental.saferExceptions
```

> Now it is possible to mark types with exception that can be thrown during the evaluation:

Reviewer: ```suggestion
Now it is possible to mark types with the type of exception that can be thrown during the evaluation:
```

### Simplified Manifest synthesis

For the sake of compatibility with Scala 2, the Scala 3 compiler now generates given instances for the `Manifest` trait. The new implementation is a slightly simplified approximation of the Scala 2 implementation. It guarantees that any of the expressions:

- `manifest[A] == manifest[B]`
- `manifest[A].runtimeClass == manifest[B].runtimeClass`
- `optManifest[A] == optManifest[B]`
- `optManifest[A].asInstanceOf[ClassTag[A]].runtimeClass == optManifest[B].asInstanceOf[ClassTag[B]].runtimeClass`

that were true in Scala 2, will also remain true when compiled with Scala 3.

### Mirrors for Hierarchical Sealed Types

Consider the following sealed hierarchy:

```scala
sealed trait Top
case class Middle() extends Top with Bottom
sealed trait Bottom extends Top
```

Previously you would not be able to summon an instance of `scala.deriving.Mirror.SumOf` for `Top`, as its child `Bottom` is not a case class. In Scala 3.1 this is now possible, enabling compiletime reflection over nested hierarchies of sealed types.

> The only change necessary for users is that you should now consider that one of the `MirroredElemTypes` for a `Mirror.SumOf` could also be a sum type, not just product types.

Reviewer: ```suggestion
The only change necessary for users is that they should now consider that one of the `MirroredElemTypes` for a `Mirror.SumOf` could also be a sum type, not just product types.
```

### Other changes

- Scastie was integrated into Scaladoc to make snippets interactive.
- `@experimental` annotation is no longer itself experimental. The users are no longer required to use a nightly build of the compiler to be able to define experimental APIs. See more detail in [the design document](https://github.com/lampepfl/dotty/blob/release-3.1.0/docs/docs/reference/other-new-features/experimental-defs.md)
- Now `TastyInspector.{inspectTastyFiles, inspectTastyFilesInJar, inspectAllTastyFiles}` return a boolean value indicating whether the process succeeded
- A `Wildcard` was made a subtype of `Ident` in the reflection API
- `TypedOrTest` was added as a supertype of `Typed` in the reflection API
- `Unapply.apply` was added to allow contruction of `Unapply` trees from macros

> - Unreductible match types are now raising type error

Reviewer: ```suggestion
- Unreductible match types now raise type errors
```

---
layout: blog-detail
post-type: blog
by: Paweł Marks, VirtusLab
title: Scala 3.1.0 released!
---
Hello from the Scala 3 team! It has already been six weeks since we have announced the release candidate for the first minor version after the initial release of Scala 3. Now, after two more RCs, we can confidently promote Scala 3.1.0-RC3 to a new stable release of the language. In other words, Scala 3.1.0 is officially out!

## Compatibility notice

> This is a first *minor* release after the initial release of Scala 3.0. This has following consequences:

Reviewer 2: perhaps insert somewhere in this section: "Scala 2 version numbers were epoch.major.minor, Scala 3 version numbers are major.minor.patch."

In my experience, it's remarkable how many people either don't know this yet, or maybe sort of vaguely remember hearing something of the sort at some point in the past, but could use a reminder.

- Scala 3.1 is backward binary compatible: you can use dependencies compiled with Scala 3.0 in 3.1 projects.
- Scala 3.1 is _not_ forward binary compatible: you _cannot_ use dependencies compiled with Scala 3.1 in Scala 3.0 projects.

> - Updating Scala to 3.1 is not a forward compatible change. If you are a library maintainer and are following SemVer you may want to update in your next minor release.

Reviewer 2: The messaging in this bullet feels different than the messaging below; the message here has a "update, but bump your minor version" feel, whereas the wording at the end of the section has more of a "don't update".

The two wordings aren't outright contradictory, exactly, but taken together, I find them a bit confusing and I think other readers will as well.

It's not the only possible solution, but I think I would suggest omitting this bullet and saying whatever needs to be said about this in the paragraph at the end of the section.

> Although we cannot guarantee full source compatibility between minor versions, we have put a lot of effort into assuring that all code that was working in 3.0.2, except some rare cases, will also work in 3.1.0. This means that if you are an application developer, you can confidently update the compiler version to take advantage of the newest improvements.

Reviewer 3: (I know this is redundant with what we said above, but I think it’s better to emphasize this point)

```suggestion
Although we cannot guarantee full source compatibility between minor versions, we have put a lot of effort into assuring that all code that was working in 3.0.2, except in some rare cases, will also work in 3.1.0. This means that if you are an application developer, you can confidently update the compiler version to take advantage of the newest improvements. You will still be able to use dependencies compiled with Scala 3.0.
```

> If you are a library maintainer, *updating to 3.1.0 will force all of your users to need to update as well*. You may consider still publishing your library using 3.0.2 to ensure that users who are reluctant to update a compiler can still use it. On the other hand, we still encourage you to test your library using 3.1.0. We understand that the current state of binary compatibility may be unsatisfactory for library maintainers. We are actively working on improvements in this area in future releases of Scala. We are taking the problem of forward-compatibility seriously, and we will provide a technical solution for 3.2.0

Reviewer 4: ```suggestion
If you are a library maintainer, *updating to 3.1.0 will force all of your users to need to update as well*. You may consider still publishing your library using 3.0.2 to ensure that users who are reluctant to update a compiler can still use it. On the other hand, we still encourage you to test your library using 3.1.0. 

*We understand that the current state of binary compatibility may be unsatisfactory for library maintainers. We are actively working on improvements in this area in future releases of Scala. We are taking the problem of forward-compatibility seriously, and we will provide a technical solution for 3.2.0*
```

> If you are a library maintainer, *updating to 3.1.0 will force all of your users to need to update as well*. You may consider still publishing your library using 3.0.2 to ensure that users who are reluctant to update a compiler can still use it. On the other hand, we still encourage you to test your library using 3.1.0. We understand that the current state of binary compatibility may be unsatisfactory for library maintainers. We are actively working on improvements in this area in future releases of Scala. We are taking the problem of forward-compatibility seriously, and we will provide a technical solution for 3.2.0

Reviewer 3: ```suggestion
If you are a library maintainer, *updating to 3.1.0 will force all of your users to update to 3.1.0 as well*. We understand that the current state of binary compatibility may be unsatisfactory. We are actively working on technical solutions to support forward-compatibility in 3.2.0. In the meantime, we recommend testing your library with Scala 3.1.0 and 3.0.2, but _publishing_ it with 3.0.2. This will allow downstream users to use your library even if they can’t update to Scala 3.1.0. You can find an example of build that follows this recommendation [here](https://github.com/typelevel/scalacheck/pull/847).
```

(not 100% sure… I think it’s nice to provide an example that people can replicate, though)

> If you are a library maintainer, *updating to 3.1.0 will force all of your users to need to update as well*. You may consider still publishing your library using 3.0.2 to ensure that users who are reluctant to update a compiler can still use it. On the other hand, we still encourage you to test your library using 3.1.0. We understand that the current state of binary compatibility may be unsatisfactory for library maintainers. We are actively working on improvements in this area in future releases of Scala. We are taking the problem of forward-compatibility seriously, and we will provide a technical solution for 3.2.0

Reviewer 5: ```suggestion
If you are a library maintainer, *updating to 3.1.0 will force all of your users to need to update as well*. You may consider still publishing your library using 3.0.2 to ensure that users who are reluctant to update a compiler can still use it. On the other hand, we still encourage you to test your library using 3.1.0. We understand that the current state of binary compatibility may be unsatisfactory for library maintainers. We are actively working on improvements in this area for future releases of Scala. We are taking the problem of forward-compatibility seriously, and we will provide a technical solution for 3.2.0
```

> If you are a library maintainer, *updating to 3.1.0 will force all of your users to update to 3.1.0 as well*. We understand that the current state of binary compatibility may be unsatisfactory. We are actively working on technical solutions to support forward-compatibility in 3.2.0. In the meantime, we recommend testing your library with Scala 3.1.0 and 3.0.2, but _publishing_ it with 3.0.2. This will allow downstream users to use your library even if they can’t update to Scala 3.1.0. You can find an example of build that follows this recommendation [here](https://github.com/typelevel/scalacheck/pull/847).

Reviewer 2: > We are actively working on technical solutions to support forward-compatibility in 3.2.0.

Is this a promise? It isn't, right? We aren't sure if this will actually happen in 3.2.0? If so, consider weakening the wording a little, perhaps to "We are exploring possible technical solutions to..."

### Other changes

- Scastie was integrated into Scaladoc to make snippets interactive.
- `@experimental` annotation is no longer itself experimental. The users are no longer required to use a nightly build of the compiler to be able to define experimental APIs. See more detail in [the design document](https://github.com/lampepfl/dotty/blob/release-3.1.0/docs/docs/reference/other-new-features/experimental-defs.md)
- Now `TastyInspector.{inspectTastyFiles, inspectTastyFilesInJar, inspectAllTastyFiles}` return a boolean value indicating whether the process succeeded
- A `Wildcard` was made a subtype of `Ident` in the reflection API
- `TypedOrTest` was added as a supertype of `Typed` in the reflection API
- `Unapply.apply` was added to allow contruction of `Unapply` trees from macros
- Unreductible match types are now raising type error
- Scala 3.1 targets Scala.js 1.7.x+. This means that users must upgrade to Scala.js 1.7.0 or later to use Scala 3.1.

Beside that Scala 3.1.0 introduced multiple small improvements and fixed a handful of bugs. You can see [the detailed changelog](https://github.com/lampepfl/dotty/releases/tag/3.1.0) on GitHub.

## What's next

> During the Scala 3.1.0 stabilization period, which took the last six weeks, we haven't stopped improving the language and fixing the in the compiler. You can already test the results of our work, using the published release candidates: Scala 3.1.1-RC1. [The full changelog](https://github.com/lampepfl/dotty/releases/tag/3.1.1-RC1) is as always available on GitHub.

Reviewer 2: "fixing the in the"? was this supposed to be "fixing bugs in the"?

Reviewer 2: also shouldn't "candidates" be "candidate"?

Thank you to all the contributors who made the release of 3.1.0 possible 🎉

According to `git shortlog -sn --no-merges 3.0.2..3.1.0` these are:

```
    61  Martin Odersky
    51  tanishiking
    50  Nicolas Stucki
    39  Kacper Korban
    33  Andrzej Ratajczak
    22  Dale Wijnand
    22  Olivier Blanvillain
    21  Lukas Rytz
    17  Filip Zybała
    16  Yichen Xu
    15  Rikito Taniguchi
    14  Sébastien Doeraene
    12  Krzysztof Romanowski
    11  EnzeXing
     9  Guillaume Martres
     9  Jamie Thompson
     8  Tom Grigg
     6  Seth Tisue
     5  Alec Theriault
     4  Michał Pałka
     4  Phil
     3  Fengyun Liu
     3  Paweł Marks
     2  Vadim Chelyshov
     2  Dmitrii Naumenko
     2  Julien Richard-Foy
     2  Kevin Lee
     2  Liu Fengyun
     2  Stéphane Micheloud
     2  Tomasz Godzik
     2  Adrien Piquerez
     2  adampauls
     2  noti0na1
     1  Kai
     1  Jeremy Smith
     1  Som Snytt
     1  Anselm von Wangenheim
     1  Boris
     1  Arnout Engelen
     1  odersky
     1  Anatolii Kmetiuk
     1  PJ Fanning
     1  Matthieu Bovel
     1  Performant Data
     1  Mario Bucev
     1  Ruslan Shevchenko
```

## Library authors: Join our community build

> Scala 3 now has a set of widely-used community libraries that are built against every nightly Scala 3 snapshot.

Reviewer 2: omit "now"
